packages/ts/react-crud/src/autogrid.tsx

```diff
@@ -130,53 +180,133 @@ type GridElementWithInternalAPI<TItem = GridDefaultItem> = GridElement<TItem> &
+        ignoreCase: false,
+      })),
+  };
+}
+
+type PageRequest = {
+  pageNumber: number;
+  pageSize: number;
+  sort: Sort;
+};
+
+type DataPage<TItem> = {
+  items: TItem[];
+  pageRequest: PageRequest;
+};
+
+async function fetchDataPage<TItem>(
+  params: GridDataProviderParams<TItem>,
   service: ListService<TItem>,
   filter: MutableRefObject<FilterUnion | undefined>,
-): GridDataProvider<TItem> {
-  let first = true;
+): Promise<DataPage<TItem>> {
+  const sort = createSort(params);
+
+  const pageNumber = params.page;
+  const { pageSize } = params;
+  const pageRequest = {
+    pageNumber,
+    pageSize,
+    sort,
+  };
+  const items = await service.list(pageRequest, filter.current);
 
+  return { items, pageRequest };
+}
+
+function createInfiniteDataProvider<TItem>(
+  service: ListService<TItem>,
+  filter: MutableRefObject<FilterUnion | undefined>,
+  grid: GridElement<TItem>,
+  afterCallBackExecuted: () => void,
+): GridDataProvider<TItem> {
   // eslint-disable-next-line @typescript-eslint/no-misused-promises
   return async (params: GridDataProviderParams<TItem>, callback: GridDataProviderCallback<TItem>) => {
-    const sort: Sort = {
-      orders: params.sortOrders
-        .filter((order) => order.direction != null)
-        .map((order) => ({
-          property: order.path,
-          direction: order.direction === 'asc' ? Direction.ASC : Direction.DESC,
-          ignoreCase: false,
-        })),
-    };
-
-    const pageNumber = params.page;
-    const { pageSize } = params;
-    const req = {
-      pageNumber,
-      pageSize,
-      sort,
-    };
-
-    const items = await service.list(req, filter.current);
-    let size;
+    const { items, pageRequest } = await fetchDataPage(params, service, filter);
+    const { pageNumber, pageSize } = pageRequest;
+    let infiniteScrollingSize;
     if (items.length === pageSize) {
-      size = (pageNumber + 1) * pageSize + 1;
-
+      infiniteScrollingSize = (pageNumber + 1) * pageSize + 1;
       const cacheSize = (grid as GridElementWithInternalAPI<TItem>)._dataProviderController.rootCache.size;
-      if (cacheSize !== undefined && size < cacheSize) {
+      if (cacheSize !== undefined && infiniteScrollingSize < cacheSize) {
         // Only allow size to grow here to avoid shrinking the size when scrolled down and sorting
-        size = undefined;
+        infiniteScrollingSize = undefined;
       }
     } else {
-      size = pageNumber * pageSize + items.length;
+      infiniteScrollingSize = pageNumber * pageSize + items.length;
     }
-    callback(items, size);
+
+    callback(items, infiniteScrollingSize);
+    afterCallBackExecuted();
+  };
+}
+
+function createFiniteDataProvider<TItem>(
+  service: CountService<TItem> & ListService<TItem>,
+  filter: MutableRefObject<FilterUnion | undefined>,
+  footerRef: MutableRefObject<Dispatch<SetStateAction<number>>>,
+  itemCountHolder: AutoGridItemCountHolder,
+  afterCallBackExecuted: () => void,
```

Comment:
Would suggest `afterPageLoaded` to give some semantic meaning.

Reply:
Done.